File: direct/test/java/ru/yandex/direct/core/testing/info/AdGroupInfo.java
package ru.yandex.direct.core.testing.info;

import java.util.Optional;

import ru.yandex.direct.core.entity.adgroup.model.AdGroup;
import ru.yandex.direct.core.entity.adgroup.model.AdGroupType;
import ru.yandex.direct.dbutil.model.ClientId;

import static ru.yandex.direct.utils.CommonUtils.ifNotNull;

public class AdGroupInfo {

    private CampaignInfo campaignInfo = new CampaignInfo();
    private AdGroup adGroup;

    public CampaignInfo getCampaignInfo() {
        return campaignInfo;
    }

    public AdGroupInfo withCampaignInfo(CampaignInfo campaignInfo) {
        this.campaignInfo = campaignInfo;
        return this;
    }

    public ClientInfo getClientInfo() {
        return getCampaignInfo().getClientInfo();
    }

    public AdGroupInfo withClientInfo(ClientInfo clientInfo) {
        getCampaignInfo().withClientInfo(clientInfo);
        return this;
    }

    public AdGroup getAdGroup() {
        return adGroup;
    }

    public AdGroupInfo withAdGroup(AdGroup adGroup) {
        this.adGroup = adGroup;
        return this;
    }

    public Long getAdGroupId() {
        return Optional.ofNullable(adGroup).map(AdGroup::getId).orElse(null);
    }

    public AdGroupType getAdGroupType() {
        return adGroup.getType();
    }

    public Long getCampaignId() {
        return ifNotNull(campaignInfo, CampaignInfo::getCampaignId);
    }

    public Long getUid() {
        return getCampaignInfo().getUid();
    }

    public ClientId getClientId() {
        return getCampaignInfo().getClientId();
    }

    public Integer getShard() {
        return getCampaignInfo().getShard();
    }

    public boolean isExists() {
        return getAdGroupId() != null;
    }
}

File: direct/test/java/ru/yandex/direct/core/testing/info/adgroup/AdGroupInfo.kt
package ru.yandex.direct.core.testing.info.adgroup

import ru.yandex.direct.core.entity.adgroup.model.AdGroup
import ru.yandex.direct.core.entity.adgroup.model.AdGroupType
import ru.yandex.direct.core.testing.info.ClientInfo
import ru.yandex.direct.core.testing.info.campaign.CampaignInfo
import ru.yandex.direct.dbutil.model.ClientId

abstract class AdGroupInfo<T : AdGroup> {
    lateinit var campaignInfo: CampaignInfo<*>
    lateinit var adGroup: T

    val shard: Int get() = campaignInfo.shard
    val uid: Long get() = campaignInfo.uid

    val clientId: ClientId? get() = campaignInfo.clientId
    val campaignId: Long get() = campaignInfo.id
    val adGroupId: Long get() = adGroup.id
    val adGroupType: AdGroupType get() = adGroup.type

    val clientInfo: ClientInfo get() = campaignInfo.clientInfo

    fun isCampaignInfoInitialized() = ::campaignInfo.isInitialized
    fun isAdGroupInitialized() = ::adGroup.isInitialized

    open fun withCampaignInfo(campaignInfo: CampaignInfo<*>) = apply { this.campaignInfo = campaignInfo }

    open fun withAdGroup(adGroup: T) = apply { this.adGroup = adGroup }
}
